feat: Split off direct-client and terminal chat from agent

reffan opened the pull request:
<!-- Use this template by filling in information and copy and pasting relevant items out of the html comments. -->

# Relates to:

Splitting off the direct-client and terminal chat from the main agent in order to achieve more control and insight. 

<!-- LINK TO ISSUE OR TICKET -->

<!-- This risks section is to be filled out before final review and merge. -->

# Risks

Medium

Users could be confused about the missing TERMINAL client without a proper docs update.

<!--
Low, medium, large. List what kind of risks, and what could be effected.
-->

# Background

## What does this PR do?

Both the direct-client start() and chat() commands usually are fired when initializing the agent. This PR makes a small refactor to the agent, bringing it more in line with how clients usually get initialized. Also split of the chat() command and related functions to the packages/client-terminal. In order to enable the in line terminal chat again the client TERMINAL must be added to the character.

## What kind of change is this?

Improvements

<!--
Bug fixes (non-breaking change which fixes an issue)
Improvements (misc. changes to existing features)
Features (non-breaking change which adds functionality)
Updates (new versions of included code)
-->

<!-- This "Why" section is most relevant if there is no linked issue explaining why. If there is a related issue it might make sense to skip this why section. -->
<!--
## Why are we doing this? Any context or related work?
-->

My reasoning for why is to bring the starting of the clients more in line with normal procedures. It also grants more control over and insight into the agents making it less of a mystery why some things work and others don't.

# Documentation changes needed?

Slight changes needed, TERMINAL should be added and could/should be added to the defaultCharacter.

<!--
My changes do not require a change to the project documentation.
My changes require a change to the project documentation.
If a docs change is needed: I have updated the documentation accordingly.
-->

<!-- Please show how you tested the PR. This will really help if the PR needs to be retested, and probably help the PR get merged quicker. -->

# Testing

## Where should a reviewer start?

## Detailed testing steps

<!--
None, automtated tests are fine.
-->

<!--
- As [anon/admin], go to [link]
  - [do action]
  - verify [result]
-->

<!-- If there is a UI change, please include before and after screenshots or videos. This will speed up PRs being merged. It is extra nice to annotate screenshots with arrows or boxes pointing out the differences. -->
<!--
## Screenshots
### Before
### After
-->

<!-- If there is anything about the deploy, please make a note. -->
<!--
# Deploy Notes
-->

<!--  Copy and paste commandline output. -->
<!--
## Database changes
-->

<!--  If there is something more than the automated steps, please specifiy deploy instructions. -->
<!--
## Deployment instructions
-->

<!-- If you are on Discord, please join https://discord.gg/ai16z and state your Discord username here for contribute role and join us in #development-feed -->
<!--
## Discord username

-->

iamladi committed:
Fix linter issues

bmgalego committed:
Merge branch 'main' into cache-manager

bmgalego committed:
wip

bmgalego committed:
wip

metadiver committed:
Create README_ES.md

Add spanish translation

lalalune committed:
Merge pull request #388 from boyaloxer/fix-tweet-truncation

Fix tweet truncation issue by truncating at complete sentences

lalalune committed:
Merge pull request #400 from metadiver/main

feat: Create README_ES.md

lalalune committed:
Merge pull request #396 from martincik/fix/register-memory-managers-of-runtime-passed-to-contructor

Register memory managers if passed to runtime

Commit:
correct relative paths DTS Build error #401

ponderingdemocritus committed:
improve logging

ponderingdemocritus committed:
Merge branch 'main' of https://github.com/ai16z/eliza into unruggable

leomercier committed:
Merge pull request #402 from ai16z/fix/tsup-build

fix: tsup build error (client-twitter)

ponderingdemocritus committed:
formatting

ponderingdemocritus committed:
contain

lalalune committed:
Merge pull request #380 from whonion/main

docs: add Russian(RU) translation of README

lalalune committed:
Merge pull request #377 from xclicx/main

docs: Update README_FR.md

lalalune committed:
Merge pull request #376 from oguzserdar/oguzserdar-patch-1

docs: add Turkish (TR) translation of README

ponderingdemocritus committed:
pnpm

ponderingdemocritus committed:
imports

ponderingdemocritus committed:
lock

ponderingdemocritus committed:
Merge pull request #383 from ai16z/ropresearch/contextual-threads

feat: Contextual Twitter Threads + Spam Reduction

ponderingdemocritus committed:
Merge branch 'main' of https://github.com/ai16z/eliza

ponderingdemocritus committed:
lint

ponderingdemocritus committed:
Merge pull request #404 from ai16z/lint

fix: Lint

ponderingdemocritus committed:
console

lalalune committed:
Merge pull request #381 from bmgalego/agent-groq

fix: adds Groq to getTokenForProvider

ponderingdemocritus committed:
blobert

ponderingdemocritus committed:
unruggable

lalalune committed:
Merge pull request #405 from ai16z/twitter-client

fix: console

ponderingdemocritus committed:
pnpm

ponderingdemocritus committed:
Merge pull request #398 from ai16z/unruggable

feat: unruggable

lalalune committed:
update readme

lalalune committed:
Merge branch 'main' of https://github.com/ai16z/eliza into HEAD

ponderingdemocritus committed:
fixes bugs

ponderingdemocritus committed:
blobert

ponderingdemocritus committed:
fixe

ponderingdemocritus committed:
Merge pull request #407 from ai16z/fixes

fix: Fixes

wahndo committed:
docs: capitalize flexible in tagline

normand1 committed:
merge from main

jkbrooks committed:
Merge pull request #389 from wahndo/main

docs: refresh eliza's tagline, fix broken links, unify formatting for core concepts

whalelephant committed:
docs: add GROK_API_KEY

lalalune committed:
Merge pull request #409 from whalelephant/main

docs: add GROK_API_KEY

normand1 committed:
update to address review feedback

normand1 committed:
inline user model

normand1 committed:
remove hardcoded endpoint

fabrizioff committed:
docs: add italian README translation

bmgalego committed:
implemented db cache adapters

ponderingdemocritus committed:
services

ponderingdemocritus committed:
remove character

ponderingdemocritus committed:
fix logging

ponderingdemocritus committed:
linting

o-on-x committed:
removed ollama embeddings. fastembeddings or openai only

jkbrooks committed:
Merge pull request #411 from fabrizioff/docs/translate-readme-it

Add italian README.md translation

dorianjanezic committed:
Add video generation plugin tests and constants

ponderingdemocritus committed:
systems work

ponderingdemocritus committed:
character

ponderingdemocritus committed:
services

RedBeardEth committed:
init deploy unruggable memecoin

Adds the deploy token action from Unruggable Memes

sirkitree committed:
Merge pull request #394 from dorianjanezic/main

feat: video generation plugin

denizekiz committed:
Readme update WSL  2 link added. (#419)

* WSL 2 instrunctions link added to readme.md

* WSL2 instructions to readme.md added

ponderingdemocritus committed:
merge

ponderingdemocritus committed:
issues

sirkitree committed:
Merge pull request #412 from ai16z/fix-services

feat: services

tarrencev committed:
Update adapters.md psql schema

I found I needed these schemas (ported from the supabase migration) in order to successfully run an agent with the discord client

sirkitree committed:
Merge pull request #413 from o-on-x/main

fix: removed ollama embeddings. fastembeddings or openai only

tarrencev committed:
Fix a few postgres issues

normand1 committed:
handle merge conflicts

ponderingdemocritus committed:
Merge pull request #424 from tarrencev/patch-1

fix: Update adapters.md psql schema

jkbrooks committed:
Merge pull request #365 from normand1/dn/token-provider-tests

feat: [Issue-185] Token Provider Tests

ponderingdemocritus committed:
Merge pull request #418 from RedBeardEth/unruggable-deploy-token

feat: unruggable on starknet

odilitime committed:
don't continue to load if a specified file is not found

odilitime committed:
don't require .env to exist

ponderingdemocritus committed:
learna an npm

odilitime committed:
add note about .env being optional

ponderingdemocritus committed:
fix build

ponderingdemocritus committed:
up

mcp97 committed:
Update contributing.md to incorporate PR Title / Description / Issue / Bug Guidelines

mcp97 committed:
Update pr.yaml to use actual check

ponderingdemocritus committed:
Merge pull request #426 from odilitime/stable-11-17

fix: don't continue to load if a specified file is not found

ponderingdemocritus committed:
Merge branch 'main' of https://github.com/ai16z/eliza into release

ponderingdemocritus committed:
Merge pull request #428 from ai16z/release

feat: lerna an npm

ponderingdemocritus committed:
v0.1.1

actions-user committed:
chore: update changelog

bmgalego committed:
Merge remote-tracking branch 'upstream/main' into cache-manager

odilitime committed:
since agent is moved out of packages, adjust default path

ponderingdemocritus committed:
configs

bmgalego committed:
remove createDirectRuntime

ponderingdemocritus committed:
config

ponderingdemocritus committed:
Merge pull request #431 from ai16z/fix-config

fix: configs

bmgalego committed:
fix

ponderingdemocritus committed:
v0.1.2

actions-user committed:
chore: update changelog

sirkitree committed:
Merge pull request #430 from monilpat/patch-1

docs: Update contributing.md to incorporate Contribution Guidelines

ponderingdemocritus committed:
linting

ponderingdemocritus committed:
Merge pull request #433 from ai16z/imports

fix: linting and imports ready for npm

ponderingdemocritus committed:
v0.1.3-alpha.0

actions-user committed:
chore: update changelog

actions-user committed:
chore: update changelog

actions-user committed:
chore: update changelog

jkbrooks committed:
Merge pull request #427 from odilitime/env-change

feat: don't require .env to exist

jkbrooks committed:
Merge pull request #429 from monilpat/main

chore: Update pr.yaml to show actual condition so easier to follow

ponderingdemocritus committed:
v0.1.3-alpha.1

actions-user committed:
chore: update changelog

ponderingdemocritus committed:
imports

ponderingdemocritus committed:
Merge pull request #435 from ai16z/fix-imports

fix: imports

ponderingdemocritus committed:
v0.1.3-alpha.2

actions-user committed:
chore: update changelog

ponderingdemocritus committed:
path

ponderingdemocritus committed:
Merge pull request #436 from ai16z/fix-paths

fix: path

ponderingdemocritus committed:
Merge pull request #432 from odilitime/fix-character-path

fix: since agent is moved out of packages, adjust default path

iamladi committed:
Fix linter issues and add console.log instead

iamladi committed:
Add console.log instead of ignore the error

jkbrooks committed:
Merge pull request #397 from martincik/fix/fix-linter-issues

fix: Fix linter issues

iamladi committed:
The bot is by default defeaned and we don't want that

o-on-x committed:
Feature Modification: Add all the stlye guidelines to the context

ponderingdemocritus committed:
fixes some console logs

lalalune committed:
Merge pull request #441 from o-on-x/main

feat:  add all the style guidelines to the context

lalalune committed:
Merge pull request #440 from ai16z/cleanups

fix: fixes some console logs

lalalune committed:
Merge pull request #437 from martincik/fix/undefean-the-bot-and-use-voice-manager

fix: The bot is by default deafened and we don't want that

odilitime committed:
Fix my name

autotranslate got it really wrong

ponderingdemocritus committed:
unrug

darwintree committed:
fix: ignored modelEndpointOverride in generation

ponderingdemocritus committed:
fix voice perms

ponderingdemocritus committed:
Merge pull request #444 from ai16z/unrug-fix

fix: unrug

ponderingdemocritus committed:
Merge pull request #447 from ai16z/fix/voice-perms

fix: voice perms

ponderingdemocritus committed:
v0.1.3

actions-user committed:
chore: update changelog

ponderingdemocritus committed:
readme

ponderingdemocritus committed:
Reade

ponderingdemocritus committed:
prettier

ponderingdemocritus committed:
Merge pull request #449 from ai16z/readme

feat: readme and linting

ponderingdemocritus committed:
Merge branch 'fewfixes' of https://github.com/tarrencev/eliza into tarrencev-fewfixes

ponderingdemocritus committed:
gitignore

ponderingdemocritus committed:
Merge branch 'tarrencev-fewfixes'

ponderingdemocritus committed:
Merge branch 'main' of https://github.com/ai16z/eliza

laser-riot committed:
fix: X dry run

bmgalego committed:
fix ensure dataDir exists

bmgalego committed:
add data folder to .gitignore

tarrencev committed:
feat: init github client

jkbrooks committed:
Merge pull request #452 from laser-riot/fix/x-dry-run

fix: X dry run

null-hax committed:
Add npm install instructions to homepage header

madjin committed:
Merge pull request #459 from null-hax/patch-1

Add npm install instructions to homepage header

madjin committed:
Merge pull request #442 from odilitime/patch-1

docs: Fix my name in stream notes

iamladi committed:
Add missing extension for levenshtein method

coffeeorgreentea committed:
feat: create-eliza-app

coffeeorgreentea committed:
chore: create-eliza-app readme

coffeeorgreentea committed:
chore: update readme package name

coffeeorgreentea committed:
chore: cleanup readme

ponderingdemocritus committed:
Merge pull request #462 from coffeeorgreentea/create-eliza-app

feat: create-eliza-app

bmgalego committed:
fix schemas add CREATE IF NOT EXISTS to tables
implement init method into db adapters

snobbee committed:
Create best-practices.md documentation

bmgalego committed:
check null values in sqlite

ai16z-demirix committed:
Fixing failing videoGeneration and token tests.

bmgalego committed:
added sqlite env config

ai16z-demirix committed:
Fixing failing videoGeneration tests.

ponderingdemocritus committed:
adds env checker for character

ponderingdemocritus committed:
character

ponderingdemocritus committed:
Merge pull request #460 from martincik/fix/fix-up-the-postgresql-schema

fix: Add missing fuzzystrmatch extension for levenshtein() method to postgresql schema.sql definition

ponderingdemocritus committed:
env in clients

ponderingdemocritus committed:
WIP

jkbrooks committed:
Merge pull request #465 from ai16z-demirix/main

fix: Fixing failling tests token.test.ts and videoGeneration.test.ts

odilitime committed:
Add Discord username question

ponderingdemocritus committed:
Merge pull request #456 from tarrencev/githubclient

feat: init github client

darwintree committed:
fix: error if openai baseURL is not provided

odilitime committed:
Update to bring inline with PR468

twilwa committed:
Merge pull request #468 from odilitime/patch-2

docs: Add Discord username question

twilwa committed:
Merge pull request #470 from odilitime/patch-3

docs: Update Contributors to bring inline with PR468

tarrencev committed:
Improve knowledge embeddings

ponderingdemocritus committed:
Merge pull request #378 from bmgalego/cache-manager

feat: Cache Manager

drew-royster committed:
ollama generate case was using console.debug. https://github.com/ai16z/eliza/issues/443

bmgalego committed:
fix ids

lalalune committed:
Merge pull request #474 from drew-royster/fix/ollama-generation-use-eliza-logger

ollama generate case was using console.debug.

ponderingdemocritus committed:
fix: ci

ponderingdemocritus committed:
Merge pull request #475 from ai16z/fix/ci

fix: ci

bmgalego committed:
use the same client for post and interactions

bmgalego committed:
fix

bmgalego committed:
save post schedule into cache

bmgalego committed:
cache new generated post

bmgalego committed:
cache profile
use ellizaLogger

bmgalego committed:
add log when creating memory

bmgalego committed:
add logs when generating text

reffan committed:
Split off the direct and terminal chat client from agent. Made terminal chat client into it's own package.

reffan committed:
Added comments to clarify.

reffan committed:
Restore cleanup flag from debug session.

haeunchin committed:
Added TWITTER_COOKIE example on quickstart.md

vivoidos committed:
add template types

bmgalego committed:
use profile, other improvements

reffan committed:
Merge branch 'main' into feature/agent-split

ponderingdemocritus committed:
Merge branch 'twitter-refactor' of https://github.com/bmgalego/eliza into bmgalego-twitter-refactor

ponderingdemocritus committed:
type

ponderingdemocritus committed:
Merge branch 'bmgalego-twitter-refactor'

lalalune committed:
Merge pull request #479 from vivoidos/template-type

refactor: add template types

ponderingdemocritus committed:
merge

ponderingdemocritus committed:
mege

bmgalego committed:
fix

bmgalego committed:
Merge branch 'ai16z:main' into twitter-refactor

bmgalego committed:
add id to char twitter profile

bmgalego committed:
Merge branch 'twitter-refactor' of https://github.com/bmgalego/eliza into twitter-refactor

ponderingdemocritus committed:
env

ponderingdemocritus committed:
revert

ponderingdemocritus committed:
Merge pull request #466 from ai16z/env

feat: adds check

ponderingdemocritus committed:
Merge pull request #446 from darwintree/patch-2

fix: ignored modelEndpointOverride in generation

bmgalego committed:
improvements

bmgalego committed:
more

bmgalego committed:
fix logs

bmgalego committed:
fix logs

bmgalego committed:
make character.username optional

ponderingdemocritus committed:
Merge pull request #482 from bmgalego/improvements

feat: Improvements

snobbee committed:
test: root test command run all packages test commands, updated best practices doc

bmgalego committed:
fix sqlite file env

bmgalego committed:
Merge branch 'improvements' of https://github.com/bmgalego/eliza into improvements

bmgalego committed:
fix direct client type

bmgalego committed:
Merge pull request #484 from bmgalego/improvements

fix: agent type error and sqlite file env

bmgalego committed:
fix: fix loadCharacters resolver

bmgalego committed:
Merge branch 'ai16z:main' into improvements

bmgalego committed:
Merge pull request #486 from bmgalego/improvements

fix: agent loadCharacters file resolver

bmgalego committed:
fix loadCharacter path gain agent does runs in root

bmgalego committed:
Merge branch 'improvements' of https://github.com/bmgalego/eliza into improvements

bmgalego committed:
Merge branch 'ai16z:main' into improvements

bmgalego committed:
Merge pull request #487 from bmgalego/improvements

fix: fix character path loading

massivefermion committed:
fix: added missing packages to tsup configs' externals

reffan committed:
Following best practices and style guides.

reffan committed:
Merge branch 'feature/agent-split' of https://github.com/reffan/eliza into feature/agent-split
Brought in line with remote repo.

jkbrooks committed:
Merge pull request #488 from massivefermion/main

fix: added missing packages to tsup configs' externals

jkbrooks committed:
Merge pull request #463 from snobbee/patch-1

docs: Create best-practices.md documentation

bmgalego committed:
Merge pull request #476 from haeunchin/document-update-for-twitter-cookie

feat: Added TWITTER_COOKIE example on quickstart.md

ponderingdemocritus committed:
Merge pull request #472 from tarrencev/main

feat: Improve knowledge embeddings

ponderingdemocritus reviewed (commented):
lets move this into the core and import what is generic into the agent directory

ponderingdemocritus committed:
improve type saftey

ponderingdemocritus committed:
Merge pull request #494 from ai16z/types

feat: improve type saftey

ponderingdemocritus committed:
improve embeddings

ponderingdemocritus committed:
revert

ponderingdemocritus committed:
improve chunks

ponderingdemocritus committed:
clean up types

ponderingdemocritus committed:
add pre-release

ponderingdemocritus committed:
Merge pull request #496 from ai16z/feat/embeddings

fix: improve embeddings

ponderingdemocritus committed:
node-v

ponderingdemocritus committed:
Merge pull request #501 from ai16z/node-v

node-v

ponderingdemocritus committed:
deps

awidearray committed:
Merge branch 'main' into dev

ponderingdemocritus committed:
Merge pull request #503 from ai16z/fix/core-deps

fix: deps

awidearray committed:
Merge pull request #502 from ai16z/dev

chore: add contributor license

actions-user committed:
chore: update changelog

ponderingdemocritus committed:
remove sol dep

ponderingdemocritus committed:
Merge pull request #504 from ai16z/solana-dep

fix: remove sol dep

ponderingdemocritus committed:
v0.1.4-alpha.0

actions-user committed:
chore: update changelog

reffan committed:
Merge branch 'main' into feature/agent-split

reffan committed:
Fixing a typo.

reffan committed:
Updated the types to add the terminal-client. Updated template character files with the new proper default clients to ensure the same user experience as before.

reffan committed:
Refactored for multi-agent mode.

reffan commented:
I believe I've fixed all of the bugs now and modified it to work in multi agent mode. Also updated the types and character files so there will be no conflict with the quickstart docs experience.
